firedrake/checkpointing.py

```diff
@@ -716,6 +720,124 @@ def _save_mesh_topology(self, tmesh):
+        ts = np.concatenate((old_ts, [t]))
+        timesteps = np.concatenate((old_timesteps, [timestep]))
+
+        # store all timesteps and indices
+        self.set_attr(path, PREFIX_TIMESTEPPING + "_indices", indices)
+        self.set_attr(path, PREFIX_TIMESTEPPING + "_times", ts)
+        self.set_attr(path, PREFIX_TIMESTEPPING + "_timesteps", timesteps)
+
+    @PETSc.Log.EventDecorator("GetTimesteps")
+    def get_timesteps(self, mesh, name):
+        """
+        Retrieve the stored indices along with associated times and timesteps for a given function.
+
+        This function is useful in checkpointing scenarios during timestepping. Each function within
+        the timestepping mode is associated with an index and, optionally, a t (time) or a timestep.
+        This method returns the stored indices, their corresponding times, and timesteps for a
+        specified function. If a function is not store in timestepping mode, empty lists are returned.
+
+        Parameters
+        ----------
+        mesh : firedrake.mesh.MeshGeometry
+            The mesh under which the function should be searched.
+        name : str
+            The name of the function for which indices, times, and timesteps are to be retrieved.
+
+        Returns
+        -------
+        indices : list
+            A list of stored indices for the function with the specified 'name'.
+        ts : list
+            A list of associated times for the indices. Each element is 'nan' if not provided for an index.
+        timesteps : list
+            A list of associated timesteps for the indices. Each element is 'nan' if not provided for an index.
+        """
+        # check if the function is mixed, or even exists in file
+        if name in self._get_mixed_function_name_mixed_function_space_name_map(mesh.name):
+            V_name = self._get_mixed_function_name_mixed_function_space_name_map(mesh.name)[name]
+            path = self._path_to_mixed_function(mesh.name, V_name, name)
+        elif name in self._get_function_name_function_space_name_map(self._get_mesh_name_topology_name_map()[mesh.name], mesh.name):
+            tmesh_name = self._get_mesh_name_topology_name_map()[mesh.name]
+            V_name = self._get_function_name_function_space_name_map(tmesh_name, mesh.name)[name]
+            path = self._path_to_function(tmesh_name, mesh.name, V_name, name)
+        else:
+            raise RuntimeError(
+                f"""Function ({name}) not found in {self.filename}""")
+
+        # check if timesteps exists for the given name and return them
+        if self.has_attr(path, PREFIX_TIMESTEPPING + "_indices"):
+            return (
+                self.get_attr(path, PREFIX_TIMESTEPPING + "_indices"),
+                self.get_attr(path, PREFIX_TIMESTEPPING + "_times"),
+                self.get_attr(path, PREFIX_TIMESTEPPING + "_timesteps"),
+                )
+        else:
+            return [], [], []
```

Comment:
```suggestion
        if name in self._get_mixed_function_name_mixed_function_space_name_map(mesh.name):
            V_name = self._get_mixed_function_name_mixed_function_space_name_map(mesh.name)[name]
            base_path = self._path_to_mixed_function(mesh.name, V_name, name)
            path = os.path.join(base_path, str(0))  # path to subfunction 0
            fsub_name = self.get_attr(path, PREFIX + "_function")
            return self.get_timesteps(mesh, fsub_name)
        elif name in self._get_function_name_function_space_name_map(self._get_mesh_name_topology_name_map()[mesh.name], mesh.name):
            tmesh_name = self._get_mesh_name_topology_name_map()[mesh.name]
            V_name = self._get_function_name_function_space_name_map(tmesh_name, mesh.name)[name]
            V = self._load_function_space(mesh, V_name)
            tV = V.topological
            path = self._path_to_function(tmesh_name, mesh.name, V_name, name)
            if PREFIX_EMBEDDED in self.h5pyfile[path]:
                path = self._path_to_function_embedded(tmesh_name, mesh.name, V_name, name)
                _name = self.get_attr(path, PREFIX_EMBEDDED + "_function")
                return self.get_timesteps(mesh, _name)
            else:
                tf_name = self.get_attr(path, PREFIX + "_vec")
                dm_name = self._get_dm_name_for_checkpointing(tV.mesh(), tV.ufl_element())
                tpath = self._path_to_vec(tV.mesh().name, dm_name, tf_name)
                if self.has_attr(tpath, PREFIX_TIMESTEPPING + "_indices"):
                    return self.get_attr(tpath, PREFIX_TIMESTEPPING + "_indices"), \
                           self.get_attr(path, PREFIX_TIMESTEPPING + "_times"), \
                           self.get_attr(path, PREFIX_TIMESTEPPING + "_timesteps")
                else:
                    return [], [], []
        else:
            raise RuntimeError(
                f"""Function ({name}) not found in {self.filename}""")
```
We probably do not need to store timestepping information directly at paths of mixed functions and embedded functions (for HDiv/HCurl) as these are just wrappers.